Shell history use k_heap instead of ring_buffer

Mattemagikern opened the pull request:
pre-requisite for: #96861

Mattemagikern committed:
shell: history rewrite to use k_heap instead of ring_buffer

The ring_buffer API was not ideal for this use case, as it is designed for
concurrent access, which is not needed here. The new implementation using
k_heap is more readable and easier to maintain.

There is a slight penalty in memory usage & effectiveness due to the
overhead of k_heap, but since this isn't a production-critical path, the
trade-off is acceptable.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

Mattemagikern committed:
shell: Remove shell history initialization function

The shell history initialization function doesn't work in if the shell
instance hasn't been created through the shell macro. This removes the
function to avoid confusion.

Signed-off-by: Måns Ansgariusson <Mansgariusson@gmail.com>

jakub-uC reviewed (changes requested):
Thank you very much for this PR — it’s a great idea to use a heap instead of a ring buffer!

Unfortunately, in its current form, this change introduces a regression in the command history functionality.
That part of the logic is quite delicate and has many corner cases. While your approach improves performance, it also causes some unexpected behavior.

**Test case**

Enter command:

`clear`→ press <_Enter_>

Enter command:
`date get`→ press <_Enter_>

(Now there are two commands in history: `clear` and `date get`)

Press the `Up` arrow repeatedly.

✅ Expected behavior (main branch)
Up   → date get
Up   → clear
Up   → [empty prompt]
Up   → date get
Up   → clear
...


(Cycles correctly through history.)

❌ Observed behavior (this PR)
Up   → date get
Up   → clear
Up   → [empty prompt]
Up   → [empty prompt]
Up/Down → still empty prompt (no access to history)

Mattemagikern commented:
Hi @jakub-uC, thanks for the review and the excellent bug-report! 

I'm fairly certain I've addressed the issue you outlined and added a test case for it.
